### DISCUSSION TITLE
Docker container and cmd options

### DISCUSSION THREAD

**Discussion by mbiuki** [schemathesis/schemathesis#1461] [Q&A] [answered]
I just found that Schemathesis has a Docker version too, which is great. I can fuzz with Schemathesis using Docker, hassle-free :-)

A few questions,

1. Do options used in CLI work for the Docker version, for example, if I have a local `openapi.yaml`, how can I pass it into the Docker container to start fuzzing it?

`docker run schemathesis/schemathesis:stable run <__local open api??__>`

2. How can I pass in the JWT token to the Docker container that runs Schemathesis?
3. How can I pass the fixed body UUIDs?
4. How can I pass data_generation_methods option: `data_generation_methods=[DataGenerationMethod.negative, DataGenerationMethod.positive]`
5. How can I pass base_url?

**Comment by mbiuki** [comment]
Additionally, I face this issue when spinning up the docker version of the Schemathesis:


```bash
mbiuki@ubuntu::~/Doc$ docker run schemathesis/schemathesis

Digest: sha256:4f1de6889806e004ffcfb4837b001a7964a94064b5943097e31ebbb59a15fe19
Status: Downloaded newer image for schemathesis/schemathesis:latest
Traceback (most recent call last):
  File "/usr/local/bin/schemathesis", line 5, in <module>
    from schemathesis.cli import schemathesis
  File "/usr/local/lib/python3.9/site-packages/schemathesis/__init__.py", line 7, in <module>
    from . import fixups, hooks, runner, serializers, targets
  File "/usr/local/lib/python3.9/site-packages/schemathesis/runner/__init__.py", line 22, in <module>
    from . import events
  File "/usr/local/lib/python3.9/site-packages/schemathesis/runner/events.py", line 8, in <module>
    from .. import DataGenerationMethod
ImportError: cannot import name 'DataGenerationMethod'
 from partially initialized module 'schemathesis' 
(most likely due to a circular import) (/usr/local/lib/python3.9/site-packages/schemathesis/__init__.py)
```

**Comment by Stranger6667** [comment, marked as answer]
Hi @mbiuki 

Sorry for not getting back to you earlier!

> Additionally, I face this issue when spinning up the docker version of the Schemathesis:

Thanks for reporting! This bug was fixed in 56bc46be205094884e750429f8e2fee3272354a7, but the Docker build failed as I was hurrying and force pushed that fix and forgot to re-run the build job. The current `latest` tag has this fix.
Though, I'd generally recommend using the `stable` tag instead:

```
docker run schemathesis/schemathesis:stable run ...
```

> Do options used in CLI work for the Docker version, for example, if I have a local openapi.yaml, how can I pass it into the Docker container to start fuzzing it?

Yep, the options are exactly the same. Though, you might want to add some Docker-specific options to work with localhost or local files. See the [latest docs](https://schemathesis.readthedocs.io/en/latest/cli.html#running-cli-via-docker) (though, they are not released yet and might slightly change):

```
docker run -v $(pwd):/mnt schemathesis/schemathesis:stable run /mnt/spec.json
```

> How can I pass in the JWT token to the Docker container that runs Schemathesis?

At the moment, the simplest option would be to use the `-H` option:

```
docker run schemathesis/schemathesis:stable run <API spec URL> -H "Authorization: Bearer <token value>"
```

When https://github.com/schemathesis/schemathesis/issues/966 is ready, it will be also possible to specify a hook for auto-refreshing auth. Though, it is also possible now with the [before_call](https://schemathesis.readthedocs.io/en/stable/extending.html#before-call) hook but requires a few extra steps:
- Write a cached version of token getter (similarly to what we discussed in #1125 )
- Use the `before_call` hook to call the token getter & update the `case` instance headers.
- Share `hooks.py` with the Docker container as in the example above
- Connect `hooks.py` with the CLI via the `--pre-run` option: `docker run schemathesis/schemathesis:stable --pre-run=hooks ...`

Let me know if I can provide more details on this.

> How can I pass the fixed body UUIDs?

I think that the easiest way would be to use the [before_generate_body](https://schemathesis.readthedocs.io/en/stable/extending.html#before-generate) hook:

```python
...

@schemathesis.hooks.register
def before_generate_body(context, strategy):

    def add_uuid(body):
         body["my-key"] = "<some uuid value>"

    return strategy.map(add_uuid)
```

And connect this hook similarly to the example in the previous block.

> How can I pass data_generation_methods option

It should be possible with the `-D` option - it could not yet be specified multiple times though :(

```
docker run schemathesis/schemathesis:stable run <API spec URL> -D negative
```

> How can I pass base_url?

With the `-b` option:

```
docker run schemathesis/schemathesis:stable run <API spec URL> -b http://test.example.com/
```

Recently I added [a full list](https://schemathesis.readthedocs.io/en/stable/cli.html#full-list-of-cli-options) of CLI options to the docs - let me know if you have any further questions! :)
